### PR TITLE
More detailed output and sorted output for "istioctl pc listeners"

### PR DESCRIPTION
**Pull request by esnible**
Resolves https://github.com/istio/istio/issues/24353

Each line is still wider than I want.

This is based on @howardjohn 's https://github.com/howardjohn/istio/blob/istioctl/listener-filterchains/istioctl/pkg/writer/envoy/configdump/listener.go .  Minor improvements:
- works with both v2 and v3 listeners
- output sorted by port and address
- MATCH column slightly more terse
- Default is `--verbose==true`

The output, on my system, is 

```
ADDRESS        PORT  MATCH                                                           DESTINATION
172.21.0.10    53    ALL                                                             outbound|53||kube-dns.kube-system.svc.cluster.local
172.21.186.46  53    ALL                                                             outbound|53||node-local-dns.kube-system.svc.cluster.local
0.0.0.0        80    ALL                                                             PassthroughCluster
0.0.0.0        80    App: HTTP                                                       80
172.21.0.1     443   ALL                                                             outbound|443||kubernetes.default.svc.cluster.local
172.21.0.219   443   ALL                                                             outbound|443||kubernetes-dashboard.kube-system.svc.cluster.local
172.21.0.219   443   App: HTTP                                                       kubernetes-dashboard.kube-system.svc.cluster.local:443
172.21.140.44  443   ALL                                                             outbound|443||istio-ingressgateway.istio-system.svc.cluster.local
172.21.187.110 443   ALL                                                             outbound|443||istiod.istio-system.svc.cluster.local
172.21.187.210 443   ALL                                                             outbound|443||public-crbq8cb78d0ojqtbm78ucg-alb1.kube-system.svc.cluster.local
172.21.232.71  443   ALL                                                             outbound|443||metrics-server.kube-system.svc.cluster.local
172.21.232.71  443   App: HTTP                                                       metrics-server.kube-system.svc.cluster.local:443
172.21.187.110 853   ALL                                                             outbound|853||istiod.istio-system.svc.cluster.local
172.21.187.110 853   App: HTTP                                                       istiod.istio-system.svc.cluster.local:853
0.0.0.0        8000  ALL                                                             PassthroughCluster
0.0.0.0        8000  App: HTTP                                                       8000
172.21.51.85   8000  ALL                                                             outbound|8000||dashboard-metrics-scraper.kube-system.svc.cluster.local
172.21.51.85   8000  App: HTTP                                                       dashboard-metrics-scraper.kube-system.svc.cluster.local:8000
0.0.0.0        8081  ALL                                                             PassthroughCluster
0.0.0.0        8081  App: HTTP                                                       8081
172.21.97.106  8081  ALL                                                             outbound|8081||catalog-operator-metrics.ibm-system.svc.cluster.local
0.0.0.0        9080  ALL                                                             PassthroughCluster
0.0.0.0        9080  App: HTTP                                                       9080
0.0.0.0        9090  ALL                                                             PassthroughCluster
0.0.0.0        9090  App: HTTP                                                       9090
172.21.0.10    9153  ALL                                                             outbound|9153||kube-dns.kube-system.svc.cluster.local
172.21.0.10    9153  App: HTTP                                                       kube-dns.kube-system.svc.cluster.local:9153
0.0.0.0        15001 ALL                                                             PassthroughCluster
0.0.0.0        15006 Addr: 172.30.206.207/32:15021                                   inbound|15021|mgmt-15021|mgmtCluster
0.0.0.0        15006 Addr: fe80::3c85:9dff:fe24:4c3d/128:9080                        Inline Route
0.0.0.0        15006 Addr: 172.30.206.207/32:9080                                    Inline Route
0.0.0.0        15006 App: Istio HTTP Plain; Addr: 172.30.206.207/32:9080             Inline Route
0.0.0.0        15006 Trans: tls; App: HTTP TLS; Addr: 0.0.0.0/0                      Inline Route
0.0.0.0        15006 App: HTTP; Addr: 0.0.0.0/0                                      Inline Route
0.0.0.0        15006 Trans: tls; App: HTTP TLS; Addr: ::0/0                          Inline Route
0.0.0.0        15006 App: HTTP; Addr: ::0/0                                          Inline Route
0.0.0.0        15006 App: Istio HTTP Plain; Addr: fe80::3c85:9dff:fe24:4c3d/128:9080 Inline Route
0.0.0.0        15006 Addr: ::0/0                                                     InboundPassthroughClusterIpv6
0.0.0.0        15006 Trans: tls; App: TCP TLS; Addr: ::0/0                           InboundPassthroughClusterIpv6
0.0.0.0        15006 Addr: 0.0.0.0/0                                                 InboundPassthroughClusterIpv4
0.0.0.0        15006 Trans: tls; App: TCP TLS; Addr: 0.0.0.0/0                       InboundPassthroughClusterIpv4
0.0.0.0        15010 ALL                                                             PassthroughCluster
0.0.0.0        15010 App: HTTP                                                       15010
172.21.187.110 15012 ALL                                                             outbound|15012||istiod.istio-system.svc.cluster.local
0.0.0.0        15014 ALL                                                             PassthroughCluster
0.0.0.0        15014 App: HTTP                                                       15014
0.0.0.0        15021 ALL                                                             Inline Route
172.21.140.44  15021 ALL                                                             outbound|15021||istio-ingressgateway.istio-system.svc.cluster.local
172.21.140.44  15021 App: HTTP                                                       istio-ingressgateway.istio-system.svc.cluster.local:15021
0.0.0.0        15090 ALL                                                             Inline Route
172.21.140.44  15443 ALL                                                             outbound|15443||istio-ingressgateway.istio-system.svc.cluster.local
0.0.0.0        50051 ALL                                                             PassthroughCluster
0.0.0.0        50051 App: HTTP                                                       50051
```

### CODE REVIEW

File: istioctl/pkg/writer/envoy/configdump/listener.go [hunk 1]
```diff
@@ -102,18 +107,166 @@ func (c *ConfigWriter) PrintListenerSummary(filter ListenerFilter) error {
...
+			filter.GetTypedConfig().TypeUrl = "type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager"
+			err := ptypes.UnmarshalAny(filter.GetTypedConfig(), httpProxy)
+			if err != nil {
+				return err.Error()
+			}
+			if httpProxy.GetRouteConfig() != nil {
+				return "Inline Route"
+			}
+			if httpProxy.GetRds().GetRouteConfigName() != "" {
+				return httpProxy.GetRds().GetRouteConfigName()
```
Comment (howardjohn) [inline istioctl/pkg/writer/envoy/configdump/listener.go:251]: maybe make this `"Route: 80"`? and for clusters, `"Cluster: outbound||blah"`?

Comment (esnible) [reply istioctl/pkg/writer/envoy/configdump/listener.go]: @howardjohn Added "Route:" and "Cluster"; also added some Match info, e.g. `Inline Route: /healthz/ready*`

**Comment by howardjohn** [comment]
what do you think of a similar flag for routes as well? to display individual routes? that one may be a bit harder, just and idea though. I'll play around with it maybe if I have time

**Comment by esnible** [comment]
> what do you think of a similar flag for routes as well? to display individual routes? that one may be a bit harder, just and idea though. I'll play around with it maybe if I have time

The "Envistion" tool had some nice CLI output of routes that could easily be brought into _istioctl_.  I haven't done it because no one has been writing issues about difficulties with `proxy-config` or complaining about it on the forums.

If I do `istioctl pc listener $POD --port 8080` I see
```
ADDRESS PORT MATCH DESTINATION
0.0.0.0 8080 ALL   Route: http.80
```

At this point it would be great to be able to do `istioctl pc route $POD --route http.80` to find out what the route looks like but _istioctl_ doesn't have a `--route` flag.  If I implemented it, would anyone use it?

**Comment by esnible** [comment]
The output of my old Enviztion tool for Route http.80 is:

- *
- path /productpage && case_sensitive true => outbound|9080||productpage.default.svc.cluster.local
(2 retries)
- prefix /static && case_sensitive true => outbound|9080||productpage.default.svc.cluster.local
(2 retries)
- path /login && case_sensitive true => outbound|9080||productpage.default.svc.cluster.local
(2 retries)
- path /logout && case_sensitive true => outbound|9080||productpage.default.svc.cluster.local
(2 retries)
- prefix /api/v1/products && case_sensitive true => outbound|9080||productpage.default.svc.cluster.local
(2 retries)

I did it that way because I thought it was the correct amount of data.  If I did it today I would replace "prefix /api/v1/products" with "/api/v1/products*" and I would remove case_sensitive true.

If this change was made consistent I we could cause `proxy-config clusters` to take a `--cluster` argument.  The output I had for a Cluster looked like this:

```
CLUSTER                      ENDPOINT
prometheus_stats        127.0.0.1:15000 healthy Success 103271
=> STATIC   
xds-grpc                        172.21.187.110:15012 healthy Error 855 Total 856
=> STRICT_DNS
Successful HTTP 2xx 856
ERRORS 503: 2
outbound_.53_._.kube-dns.kube-system.svc.cluster.local
=> EDS
```

**Pull request merged**